### PR TITLE
Fix linker file for RISCV-64 targets

### LINKED ISSUE #196: `riscv-rt`: Broken eh_frame relocations on QEMU
Demo repo: https://github.com/dreiss/panic_repro .  It's a fairly simple riscv-rt example targeting riscv64 on qemu.

On "rustc 1.78.0-nightly (a84bb95a1 2024-02-13)" or later, the link fails with bad relocations in eh_records:

```
  = note: rust-lld: error: <internal>:(.eh_frame+0x1c): relocation R_RISCV_32_PCREL out of range: 2147489994 is not in [-2147483648, 2147483647]; references ''
          >>> defined in /home/dreiss/.rustup/toolchains/nightly-2024-02-14-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-none-elf/lib/libcore-2678f83f395a7a3f.rlib(core-2678f83f395a7a3f.core.c4f53ede227ad4d2-cgu.0.rcgu.o)
          
          rust-lld: error: <internal>:(.eh_frame+0x30): relocation R_RISCV_32_PCREL out of range: 2147489980 is not in [-2147483648, 2147483647]; references ''
          >>> defined in /home/dreiss/.rustup/toolchains/nightly-2024-02-14-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-none-elf/lib/libcore-2678f83f395a7a3f.rlib(core-2678f83f395a7a3f.core.c4f53ede227ad4d2-cgu.0.rcgu.o)

...
          rust-lld: error: too many errors emitted, stopping now (use --error-limit=0 to see all errors)
```

On "rustc 1.78.0-nightly (b381d3ab2 2024-02-12)" or newer, the link succeeds, and debug_frame records for functions in my code are fine, but eh_frame records for functions in libcore are pointing to addresses around `0xffffffff8000ABCD` (confirmed with `objdump -WFL` and `llvm-readelf --unwind`).  This prevents gdb from backtracing up through libcore functions in some cases.

I think the problem is that riscv-rt's linker script is marking the eh_frame section as "(INFO)", which makes it non-allocatable, which puts it at a logical address of 0.  Which means any from it to the code (of which there are, of course, many) are (potentially) longer than 2GB, which is not allowed.  The new rustc correctly fails.  The old one (which is also an older llvm) messes up the relocation due to overflow.

Making eh_frame not defined as "(INFO)" should fix this, but then it will be included in the output binary, which is not ideal for small targets.  If unwinding data is not going to be used at runtime, it's probably best to use debug_frame instead of eh_frame, but that cannot be controlled for libcore unless it's being built from source.

### PR DESCRIPTION
**Pull request by romancardenas**
Looks like the current linker file contains some things that were probably required when using binary blobs in `riscv-rt`. Now that we use inline assembly, they are not needed. Furthermore, some of them appear to provoke errors.

This PR (still WIP) aims to fix these issues. For instance, just commenting out the `eh_frame` part of the linker script fixes the compilation of the example provided in #196 

I still need to play around with this, but this is a good first step

Closes #196

### CODE REVIEW

**Comment by romancardenas** [comment]
I am currently going through `cortex-m-rt` linker file and checking the differences. I will also assess the size of binaries etc. to evaluate the side effects of any change.

File: riscv-rt/link.x.in [hunk 1]
```diff
@@ -170,8 +170,8 @@ SECTIONS
     KEEP(*(.got .got.*));
   }
 
-  .eh_frame (INFO) : { KEEP(*(.eh_frame)) }
-  .eh_frame_hdr (INFO) : { *(.eh_frame_hdr) }
+  /* .eh_frame (INFO) : { KEEP(*(.eh_frame)) } */
+  /* .eh_frame_hdr (INFO) : { *(.eh_frame_hdr) } */
```
Comment (rmsyn) [inline riscv-rt/link.x.in:174]:
```suggestion
  .eh_frame : { KEEP(*(.eh_frame)) } > REGION_TEXT
  .eh_frame_hdr (INFO) : { *(.eh_frame_hdr) } > REGION_TEXT
```

This is the suggestion from the linked issue, and also works to resolve the issue.

I think the main problem is that The `INFO` attribute clears the `SHF_ALLOC` bit in `sh_flags`: <https://lld.llvm.org/ELF/linker_script.html#output-section-type>

The `INFO` attribute is what looks to have been causing the error, since it was telling `rust-lld` to not allocate space for the `eh_frame` and `eh_frame_hdr` sections. Then, when referenced by another part of code, they were seen as being outside the allocated memory.

So, removing the `eh_frame` and `eh_frame_hdr` sections entirely, or removing the `INFO` section resolves the issue.

Completely removing the sections may be the better option, as it allows downstream users to define them when necessary, without forcing them to place them in the `REGION_TEXT` region.

**Comment by romancardenas** [comment]
Also, `cortex-m-rt` includes an uninit section at the bottom of the RAM. We don't add anything after the stack

**Comment by romancardenas** [comment]
I suggest adopting part of `cortex-m-rt` structure:

# Make `link.x` include `memory.x`

In this way, PACs and BSPs would have to provide a `memory.x` linker file that defines the memory layout and regions.

End users would need to do something like `RUSTFLAGS="-C link-arg=-Tlink.x` when building binaries, regardless of the final target.

# Keep using region aliases

I think the current `riscv-rt` approach is quite versatile and I don't see any advantage in limiting sections to FLASH and RAM.

# Use `__` to define start and end of sections

I think it is a good idea to follow the same naming convention as in `cortex-m-rt` (`__stext`, `__etext`...).

# Add a `REGION_UNINIT`after the stack

By default, the section would be empty and everything would work as it is. However, it could be useful in some use cases

**Comment by romancardenas** [comment]
For the new release of `riscv-rt`, I would focus on the current changes in the linker file. We can explore other changes for a next iteration, if we think they are valuable

File: riscv-rt/src/lib.rs [hunk 2]
```diff
@@ -177,72 +123,189 @@
 //! REGION_ALIAS("REGION_DATA", RAM);
 //! REGION_ALIAS("REGION_BSS", RAM);
 //! REGION_ALIAS("REGION_HEAP", RAM);
-//! REGION_ALIAS("REGION_STACK", L2_LIM);
+//! REGION_ALIAS("REGION_STACK", RAM);
```
Comment (rmsyn) [inline riscv-rt/src/lib.rs:126]: `L2_LIM` is a region distinct from main ram. At least on SiFive U74 cores, this is the `SRAM` region.

File: riscv-rt/src/lib.rs [hunk 3]
```diff
@@ -1,126 +1,44 @@
-//! Minimal startup / runtime for RISC-V CPU's
+//! Startup code and minimal runtime for RISC-V microcontrollers
```
Comment (rmsyn) [inline riscv-rt/src/lib.rs:1]: Do these changes imply incompatibility with application SoCs with multiple HARTs?

Why the change from `CPU` to `microcontroller`?

Comment (romancardenas) [reply riscv-rt/src/lib.rs:1]: I just copy/pasted the docs in `cortex-m-rt`, the liker should present the exact same configuration as before

File: riscv-rt/src/lib.rs (hunk 2 above)
Comment (romancardenas) [reply riscv-rt/src/lib.rs:126]: I can rescue the previous example in which `L2_LIM` is used.

File: riscv-rt/link.x.in [hunk 4]
```diff
@@ -216,11 +233,11 @@ ASSERT(SIZEOF(.stack) > (_max_hart_id + 1) * _hart_stack_size, "
...
 ASSERT(SIZEOF(.got) == 0, "
-.got section detected in the input files. Dynamic relocations are not
-supported. If you are linking to C code compiled using the `gcc` crate
-then modify your build script to compile the C code _without_ the
--fPIC flag. See the documentation of the `gcc::Config.fpic` method for
-details.");
+ERROR(riscv-rt): .got section detected in the input files. Dynamic relocations are not
+supported. If you are linking to C code compiled using the `cc` crate then modify your
+build script to compile the C code _without_ the -fPIC flag. See the documentation of
+the `cc::Build.pic` method for details.");
```
Comment (rmsyn) [inline riscv-rt/link.x.in:241]: Would it be possible to support PIC relocation in the future? If it is, this is something I would be interested in working on.

**Review by rmsyn — APPROVED** [review]
LGTM

File: riscv-rt/link.x.in (hunk 4 above)
Comment (romancardenas) [reply riscv-rt/link.x.in:241]: It is an interesting research line. Especially for RISC-V 64 targets which currently face issues with static linking (see #153). However, it would need specific code to deal with dynamic linking etc.

**Comment by romancardenas** [comment]
Closes #196
